pkg/generator/ignition/generator.go

```diff
@@ -137,3 +154,116 @@ func ignitionFromOperatingSystemConfig(config *generator.OperatingSystemConfig)
 
 	return json.Marshal(outCfg)
 }
+
+// decodeProviderConfig decodes the provider config into the given struct
+func decodeProviderConfig(providerConfig *runtime.RawExtension, into runtime.Object) error {
+	if providerConfig == nil {
+		return nil
+	}
+
+	if _, _, err := getGardenerDecoder().Decode(providerConfig.Raw, nil, into); err != nil {
+		return fmt.Errorf("could not decode provider config: %w", err)
+	}
+
+	return nil
+}
+
+var (
+	decoder runtime.Decoder
+)
+
+// getGardenerDecoder returns a decoder to decode Gardener objects
+func getGardenerDecoder() runtime.Decoder {
+	if decoder == nil {
+		scheme := runtime.NewScheme()
+		utilruntime.Must(gardenv1beta1.AddToScheme(scheme))
+		decoder = serializer.NewCodecFactory(scheme).UniversalDecoder()
+	}
+	return decoder
+}
+
+func additionalContainterdConfigFile(mirrors []metalextensionv1alpha1.RegistryMirror) types.File {
+	containerdConfig := types.File{
+		Path:       "/etc/containerd/config.toml",
+		Filesystem: "root",
+		Mode:       &types.DefaultFileMode,
+		Contents: types.FileContents{
+			Inline: containerdBaselineConfig,
+		},
+	}
+	if len(mirrors) == 0 {
+		return containerdConfig
+	}
+
+	renderedContent := `# Generated by os-extension-metal
+imports = ["/etc/containerd/conf.d/*.toml"]
+version = 2
+
+[plugins."io.containerd.grpc.v1.cri".registry]
+  [plugins."io.containerd.grpc.v1.cri".registry.mirrors]
+`
+	for _, m := range mirrors {
+		for _, of := range m.MirrorOf {
+			renderedContent += fmt.Sprintf(`    [plugins."io.containerd.grpc.v1.cri".registry.mirrors.%q]
+      endpoint = [%q]
+`, of, m.Endpoint)
+		}
+	}
+
+	containerdConfig.Contents.Inline = renderedContent
+	return containerdConfig
+}
+
+func additionalDNSConfFiles(dnsServers []string) []types.File {
+	if len(dnsServers) == 0 {
+		return nil
+	}
+	resolveDNS := strings.Join(dnsServers, " ")
+	systemdResolvedConfd := fmt.Sprintf(`# Generated by os-extension-metal
+
+[Resolve]
+DNS=%s
+Domain=~.
+
+`, resolveDNS)
+	resolvConf := "# Generated by gardener-extension-provider-metal\n"
+	for _, ip := range dnsServers {
+		resolvConf += fmt.Sprintf("nameserver %s\n", ip)
+	}
+
```

Comment:
According to https://man7.org/linux/man-pages/man5/resolv.conf.5.html up to 3 nameservers my be listed

> Up to MAXNS (currently 3, see <resolv.h>) name servers may be listed, one per keyword.

what to do, if there are more than 3 in the `dnsServers` slice?

Reply:
Good point! This should ideally be caught by a validating webhook in the gepm.

Reply:
We should validate this in the gepm networkisolatio validation and prohibit more than 3

Reply:
Validation here: https://github.com/metal-stack/gardener-extension-provider-metal/pull/361/commits/ce898b7700dbf8883b7685509310ab5151816160